File: scripts/favicon_build.py
"""สร้าง favicon สี่เหลี่ยมจัตุรัสจาก img/โลโก้เว็บ.png — ครอปให้โลโก้ใหญ่ชัด ไม่บีบสัดส่วน."""
from pathlib import Path

try:
    from PIL import Image
except ImportError:
    raise SystemExit("ติดตั้ง Pillow ก่อน: py -m pip install pillow")

root = Path(__file__).resolve().parent.parent
src = root / "img" / "โลโก้เว็บ.png"
if not src.is_file():
    raise SystemExit(f"ไม่พบไฟล์: {src}")

img = Image.open(src).convert("RGBA")
bbox = img.getbbox()
if bbox:
    img = img.crop(bbox)
w, h = img.size
print(f"Source (trimmed): {w}x{h}")

# ครอปเป็นสี่เหลี่ยมจัตุรัสกลางภาพ แล้วขยายเต็ม canvas
side = min(w, h)
left = (w - side) // 2
top = (h - side) // 2
square = img.crop((left, top, left + side, top + side))

FILL = 0.96  # โลโก้เกือบเต็มกรอบ favicon

for size in (32, 48, 64, 128, 192):
    canvas = Image.new("RGBA", (size, size), (0, 0, 0, 0))
    inner = int(round(size * FILL))
    resized = square.resize((inner, inner), Image.Resampling.LANCZOS)
    offset = (size - inner) // 2
    canvas.paste(resized, (offset, offset), resized)
    out = root / "img" / f"favicon-{size}x{size}.png"
    canvas.save(out, optimize=True)
    print(f"Wrote {out}")

print("Done.")

File: scripts/favicon_build.php
<?php
/**
 * สร้าง favicon สี่เหลี่ยมจัตุรัสจาก โลโก้เว็บ.png (ไม่บีบสัดส่วน)
 * รัน: php scripts/favicon_build.php
 */
declare(strict_types=1);

$root = dirname(__DIR__);
$src = $root . '/img/โลโก้เว็บ.png';
if (!is_file($src)) {
    fwrite(STDERR, "Missing source: {$src}\n");
    exit(1);
}

if (!function_exists('imagecreatefrompng')) {
    fwrite(STDERR, "PHP GD extension required.\n");
    exit(1);
}

$img = @imagecreatefrompng($src);
if ($img === false) {
    fwrite(STDERR, "Cannot load PNG.\n");
    exit(1);
}

$w = imagesx($img);
$h = imagesy($img);
echo "Source: {$w}x{$h}\n";

// ครอปสี่เหลี่ยมจัตุรัสกลาง (ให้โลโก้ใหญ่ในแท็บ)
$side = min($w, $h);
$left = (int) floor(($w - $side) / 2);
$top = (int) floor(($h - $side) / 2);
$crop = imagecreatetruecolor($side, $side);
imagealphablending($crop, false);
imagesavealpha($crop, true);
$transparent = imagecolorallocatealpha($crop, 0, 0, 0, 127);
imagefilledrectangle($crop, 0, 0, $side, $side, $transparent);
imagecopy($crop, $img, 0, 0, $left, $top, $side, $side);
imagedestroy($img);
$img = $crop;
$w = $side;
$h = $side;

$fill = 0.96;
$sizes = [32, 48, 64, 128, 192];
foreach ($sizes as $size) {
    $out = imagecreatetruecolor($size, $size);
    imagealphablending($out, false);
    imagesavealpha($out, true);
    $transparent = imagecolorallocatealpha($out, 0, 0, 0, 127);
    imagefilledrectangle($out, 0, 0, $size, $size, $transparent);
    imagealphablending($out, true);

    $inner = (int) round($size * $fill);
    $dx = (int) round(($size - $inner) / 2);
    $dy = $dx;

    imagecopyresampled($out, $img, $dx, $dy, 0, 0, $inner, $inner, $w, $h);

    $path = $root . '/img/favicon-' . $size . 'x' . $size . '.png';
    imagepng($out, $path, 9);
    imagedestroy($out);
    echo "Wrote {$path}\n";
}

imagedestroy($img);
echo "Done.\n";
